docs/source/development/eep-02-typing.md

```diff
@@ -0,0 +1,1474 @@
+scalar, least-squares and likelihood optimizers. Moreover, a criterion function can
+return additional data that is stored in the log file (if logging is active). All of
+this is achieved by returning a dictionary instead of just a scalar float.
+
+The conventions for the return value of the criterion function are as follows:
+
+- For the simplest case, where only scalar optimizers are used, `criterion` returns a
+  float or a dictionary containing the entry "value" which is a float.
+- For likelihood functions, `criterion` returns a dictionary that contains at least the
+  entry `"contributions"` which can be any pytree that can be flattened into a numpy
+  array of floats.
+- For least-squares problems, `criterion` returns a dictionary that contains at least
+  the entry `"root_contributions"` which can be any pytree that can be flattened into a
+  numpy array of floats.
+- In any case the returned dictionary can have an arbitrary number of additional keys.
+  The corresponding information will be stored in the log database (if logging is used).
+
+A few simple examples of valid criterion functions are:
+
+```python
+def sphere(params: np.ndarray) -> float:
+    return params @ params
+
+
+def dict_sphere(params: dict) -> float:
+    return params["a"] ** 2 + params["b"] ** 2
+
+
+def least_squares_sphere(params: np.ndarray) -> dict[str, Any]:
+    out = {"root_contributions": params, "p_mean": params.mean(), "p_std": params.std()}
+    return out
+```
+
+**Things we want to keep**
+
+- The fact that `params` can be arbitrary pytrees makes estimagic flexible and popular.
+  We do not need to restrict this type in any way because flattening the pytree gives us
+  a very precise type no matter how complicated the tree was.
+- Using the same criterion function for scalar, likelihood and least-squares optimizers
+  makes it easy to try out and compare very different algorithms with minimal code
+  changes.
+- We do not need to restrict the type of additional arguments of the criterion function.
+- The simplest form of our criterion functions is also compatible with scipy.optimize
+
+**Problems**
+
+- Most users of estimagic find it hard to write criterion functions that return the
+  correct dictionary. Therefore, they don't use the logging feature and we often get
+  questions about specifying least-squares problems correctly.
+- Internally we can make almost no assumptions about the output of a criterion function,
+  making the code that processes the criterion output very complex and full of if
+  conditions.
+- The best typing information we could get for the output of the criterion function is
+  `float | dict[str, Any]`, which is not very useful.
+- We only know if the specified criterion function is compatible with the selected
+  optimizer after we evaluate it once. This means that errors for users are raised very
+  late.
+- While optional, in least-squares problems it is possible that a user specifies
```

Comment:
I don't get this point. Seems like a consequence of wanting to use the same criterion in multiple optimizers, which you cite as a strength above? So what precisely is the problem / what do you mean with "compatible"?

Reply:
The thing is that in all cases it would be sufficient to specify the output at the least aggregated level. For a leas-squares problem, only the `"root_contributions"` are necessary. estimagic can then calculate the `"contributions"` and `"value"` from the `"root_contributions"`. 

However users can also provide all three dictionary entries. If they do that, we check that their calculations coincide with ours. This is what I meant by compatible but I should have called it consistent.

The problem does not come from having criterion functions that are compatible with multiple classes of optimizers but from using a dictionary as signaling device for the output type. 

This is why the preferred way of specifying least-squares problems will be:

```python
@em.mark.least_squares 
def ls_sphere(params):
    return params
```

Reply:
Done.